1-Draft/RFCNNNN-Multi-Line-Continuance.md

```diff
@@ -0,0 +1,234 @@
+can be seen [here](https://sqldbawithabeard.com/2018/03/11/easily-splatting-powershell-with-vs-code/), once you've converted to splatting you still lose Intellisense
+for future updates unless you work from the command first and then add to your
+splatted collection, and that's just in Visual Studio Code. Other editors may
+or may not support that functionality, and users working in a standalone
+terminal won't have that available to them either.
+
+Instead, why not allow users to wrap commands across multiple lines in a more
+intuitive way without having to deal with backticks on every line or splatting,
+like this:
+
+```PowerShell
+New-ADUser @
+    -Name 'Jack Robinson'
+    -GivenName 'Jack'
+    -Surname 'Robinson'
+    -SamAccountName 'J.Robinson'
+    -UserPrincipalName 'J.Robinson@enterprise.com'
+    -Path 'OU=Users,DC=enterprise,DC=com'
+    -AccountPassword (Read-Host -AsSecureString 'Input Password')
+    -Enabled $true
+
+Get-ChildItem @
+    $rootFolder
+    -File
+    -Filter '*.ps*1'
+
+```
+
+Of course, they could invoke external commands and pass through arguments this
+way as well:
+
+```PowerShell
+& "./plink.exe" @
+    --% $Hostname -l $Username -pw $Password $Command
+
+cacls @
+    c:\docs\work
+    /E /T /C /G
+    "FinanceUsers":F
+
+```
+
+Further, by generalizing multi-line continuance with a `@` character, we're
+allowing users to apply line continuance the way they want to, which opens the
+door to more C#-like line wrapping when you're working with multiple members or
+methods in .NET, one after another. For example, this would work:
+
+```PowerShell
+$string @
+    .ToUpper()
+    .Trim()
+    .Length
+```
+
+In each of these examples, the parser starts parsing the command as a
+multi-line command when it encounters the `@` token as the last token on the
+line, and in this mode command parsing stops once one of the following is
+found:
+
+* end of file
+* two newlines (as opposed to the normal one)
+* command-terminating token (i.e. all other ways of ending commands work the
+same as usual, and this does not affect other elements of the PowerShell
+syntax)
+
+The pros/cons to this new syntax are as follows:
+
+**Pros:**
+
+* allows the scripter to wrap commands how they see fit, while still getting
+Intellisense and tab completion, without using backticks.
+* aside from the `@` character to initiate multi-line continuance, the rest of
+the command is entered the exact same way it would be if it was entered on a
+single line.
+* ad hoc could support this syntax as well (PSReadline could wait for a
+double-enter when in multi-line command parsing mode)
+* no breaking changes (a standalone `@` is currently an unrecognized token in
+PowerShell no matter where it is used).
+* users can use a blank line to terminate the command, or they can opt to use
+any valid command-terminating token instead, so it has a proper closing
+character.
+
+**Cons:**
+
+* no known cons at this time
```

Comment:
Well, adding "multi-line commands" as a new concept to the language might be considered a con by some. It will be unusual, and wrapping lines without a marker _on the end of each line_ might be unique to PowerShell.

You're also using up one of the few symbols that aren't currently valid syntax 😛

Reply:
For what it's worth. I still like this idea better than the `@@{` or `-@{` in-line splatting.

Reply:
> Well, adding "multi-line commands" as a new concept to the language might be considered a con by some. It will be unusual, and wrapping lines without a marker on the end of each line might be unique to PowerShell.

Sometimes being unique makes something stand out, hopefully in a positive way. The inline splatting syntax is also unusual and unique, but definitely not in a good way from my perspective.

> You're also using up one of the few symbols that aren't currently valid syntax 😛

On the symbol note, I think we're luck that the `@` symbol is invalid syntax when used like this ATM, because of the natural association it has with splatting.

> For what it's worth. I still like this idea better than the @@{ or -@{ in-line splatting.

Thanks. Me too.

Reply:
Also @Jaykul, I noticed you marked another comment with the confused emoji, which made me re-read that section and see that the way I explained it was confusing. I've since updated that section. If you're still confused by it, let me know.

Reply:
The only real con that I can see is that introducing a new language feature like this will require people being made aware of the fact this only works in newer versions of PowerShell, effectively adding further breaking functionality between versions, albeit a positive break, much like the addition of DSC or Classes to the language.

This in itself is more of an issue with language growth than this particular RFC, but I think that we need to be weary of the additional adoption piece & what needs to happen here as part of the wider v7 release, which I think needs to have some direction from @joeyaiello & @SteveL-MSFT & team on what they think needs to happen here, not only to inform and educate users in how to use this new functionality, but what they think is the path to a successful v7 adoption and how the community can get on board, with items like suggestions for new PSScriptAnalyzer rules that may help with the authoring experience, streaming/presenting/blogging about how to use new language features etc, though I will raise an issue to discuss possible enhancements to the RFC template to accommodate this going forward as I think that this will be an important aspect to consider in future.

FWIW, I think this is a much needed language addition which will simplify both the authoring experience and the readability/maintenance experience, however I do feel that to state there's `no known cons at this time` is technically incorrect, and whilst all Language features have an **implicit** breaking change in language functionality meaning that those that adopt this new feature can no longer back port scripts to prior versions, which whilst we all know this is the case I can't help but feel this needs to be **explicitly** mentioned as a con for anyone reading this coming completely fresh to PowerShell in 2/3 years with no experience in future.

I will however look forward to being able to use this functionality if/when this makes it into PowerShell.

Reply:
>...effectively adding further breaking functionality between versions, albeit a positive break...

Never, ever describe language enhancements as "adding breaking functionality". Doing so only contributes to guiding the community in a very bad direction.

Language enhancements are forward compatible enhancements to PowerShell. They are not backward compatible.

That's it.

There should be no mention of "breaking" functionality (a very negative, bad thing to be avoided) in a discussion about language enhancements (a very positive thing).

> ...I think that we need to be weary of the additional adoption piece & what needs to happen here as part of the wider v7 release...

PowerShell 7 can be installed side-by-side with downlevel versions of Windows PowerShell, and it uses a different executable name than Windows PowerShell.  That presents a much lower barrier to entry than any of the previous Windows PowerShell upgrades. The more language enhancements, new features, and performance enhancements that we have that add tangible value to users, while maintaining backwards compatibilty with existing scripts as much as possible, the more likely users will want to adopt the new version.

In general though, as a community we need to push using `#requires -Version` in scripts and setting `PowerShellVersion` in module manifests as a best practice. @kilasuit, there is another RFC related to that, at least from a module perspective, [here](https://github.com/PowerShell/PowerShell-RFC/pull/188) (be sure to read the comments/thoughts before adding them so that you understand where I'm coming from on that one).

Reply:
> Never, ever describe language enhancements as "adding breaking functionality". Doing so only contributes to guiding the community in a very bad direction.

I don't think this wording is helpful to the discussion. This is a Request for Comments, not a Defense from Comments, and the purpose of the discussion is to collaborate on design. I don't think strongly worded interdictions on technical opinion bring us closer to that objective. In fact, I think part of the point of PowerShell is to be dynamic enough as to eschew a hard orthodoxy. I'm just a developer, but I spend a lot of my time trying to soften discussions and prevent flame wars on GitHub now (time which I could spend enhancing PowerShell, etc.), and would prefer if we could avoid that.

On the note of language features generally, I think part of the consideration for each one should include the fact that any new syntax will prevent any script using it from working on downlevel PowerShell versions. This is particularly pertinent I think because the big promise of PowerShell 7 is for Windows PowerShell 5.1 compatibility.

I think a lot of discussion that happens in the PowerShell development community comes from PowerShell users who also write a lot of C#, and C# has made a name for itself pushing out great language features. But in C#, the parser is separate from the interpreter, and developers writing the C# can just update their compiler to use new language features that will work on existing IL interpreters in environments they can't update. But in PowerShell, the parser is part of the interpreter, and adding a new language syntax is just like adding new IL features (for which the bar is much much higher).

There are existing PowerShell interpreters out there; downlevel PowerShell versions, plus some syntactic tools. And just like with .NET, plenty of people don't have control over the interpreter running on their target platform. That's part of our compatibility story; not just "my PowerShell 5.1 scripts will run in PowerShell 7" but "I can continue writing scripts that will work in both PowerShell 5.1 and PowerShell 7".

Language features are certainly exciting and can be implemented in the compiler so that they theoretically run efficiently, but I think it's worth keeping in mind:
- New syntax doesn't just pose a challenge for old PowerShell versions (script and script module authors being affected), but also for the PowerShell SDK; code that targets the PowerShell AST must not crash because of changes to {I}AstVisitor{2} APIs.
- Implementing language features comes at an opportunity cost. E.g. we could add a keyword to make a particular scenario fast, or we could devote resources to performance improvements in command binding and pipeline implementation across the board.
- Language features are *difficult to implement*. They can touch the tokeniser, parser, compiler, semantic checker, completions and a whole long tail of things. They're also hard to verify and test properly because they can have all kinds of edge cases. And changes introduced with one can have far-reaching consequences; a small grammar change can unintentionally allow bad constructions, small performance changes in the tokeniser or parser will be felt everywhere.
- Because of their complexity, language features cost more to own and maintain compared to modules or cmdlets or even engine features/APIs.
- PowerShell is intended for dynamism. It's why we have modules and dynamic loading and string templating. So we should prefer a dynamic approach when we can.

Ultimately we want language features of course -- they improve PowerShell. But there are pros and cons we should weigh appropriately. Personally I think that the bar should be high because of the complexity and opportunity cost invovled.

Reply:
> I don't think this wording is helpful to the discussion. This is a Request for Comments, not a Defense from Comments, and the purpose of the discussion is to collaborate on design...

Ironically, you're saying wording is important, which is exactly what I was saying, but it was in haste as I was on my way out of the office at the time so I didn't take time to review/flush out my reply like I usually do.

What I should have said:

This RFC is not breaking. Suggesting that it effectively adds "breaking functionality" is incorrect. It extends the language, and is therefore forward compatible.

Since it seems that the wording I used came across as offensive, I apologize for that.

I do feel very strongly that people of influence should be careful about the wording used to discuss changes with the community, especially when talking about "breaking" functionality, which is considered very negative when it comes to software. New language features are not breaking. They are not backward compatible, because they are new language features, but they are absolutely not breaking. That is the wrong terminology to use, and I think it is important to call that out, especially in an RFC where others reading the RFC may also read the comments. This RFC proposal does not include any breaking changes.

> On the note of language features generally, I think part of the consideration for each one should include the fact that any new syntax will prevent any script using it from working on downlevel PowerShell versions.

It seems like overkill to me. Adding a new keyword or operator? Of course those won't work in downlevel versions. If we're going to go that route, maybe we should have two different templates for RFCs: one for internal enhancements that are forward compatible (language extensions like operators, keywords, etc.) and one for external enhancements that are backward compatible (modules), where it is clear from the template itself what you're looking at, so that RFC authors don't need to call out the obvious.

> This is particularly pertinent I think because the big promise of PowerShell 7 is for Windows PowerShell 5.1 compatibility.

Backwards compatibility. That doesn't mean that scripts leveraging PowerShell 7 language features will be compatible with Windows PowerShell 5.1. That's not part of the big promise.

Should the promise of backwards compatibility hold back future improvements in PowerShell? It didn't between versions 1, 2, 3, 4, and 5 of Windows PowerShell. It didn't with PowerShell Core 6 either. PowerShell momentum has slowed down a lot in recent years. Don't get me wrong, moving to .NET Core/cross platform is huge. But there are still so many opportunities for improvement in the core product itself that seem to take forever to see the light of day, and at the same time there is still too much attention/effort from the core team placed on releasing downlevel support, especially when PowerShell 7 will install side-by-side with Windows PowerShell.

> ...in PowerShell, the parser is part of the interpreter, and adding a new language syntax is just like adding new IL features (for which the bar is much much higher).
> 
> There are existing PowerShell interpreters out there; downlevel PowerShell versions, plus some syntactic tools. And just like with .NET, plenty of people don't have control over the interpreter running on their target platform. That's part of our compatibility story; not just "my PowerShell 5.1 scripts will run in PowerShell 7" but "I can continue writing scripts and modules that will work in both PowerShell 5.1 and PowerShell 7".

Agreed. Users who are using newer versions should still be able to create scripts or modules that work on downlevel versions. That need does not take away from the need to move the language forward with new value, however, especially since scripts and modules can specify the minimum version of PowerShell that they support. Of course, that's only useful if script and module authors actually do that, but that's a different issue/challenge to solve. The fact that issue/challenge exists shouldn't get in the way of pushing the language forward for the better.

> Language features are certainly exciting and can be implemented in the compiler so that they theoretically run efficiently, but I think it's worth keeping in mind:
> 
>     * New syntax doesn't just pose a challenge for old PowerShell versions (script and script module authors being affected), but also for the PowerShell SDK; code that targets the PowerShell AST must not crash because of changes to {I}AstVisitor{2} APIs.

Shouldn't we add specific Pester tests to catch such things?

>     * PowerShell is intended for dynamism. It's why we have modules and dynamic loading and string templating. So we should prefer a dynamic approach when we can.

This RFC is about meeting the dynamic needs of the PowerShell community when it comes to how they format their scripts for easier authoring, review and maintenance. I know that's not the point you were making, but that's what this RFC is about, just the same.

Reply:
Sorry. As I have before, I simply disagree.

This is a gratuitous breaking change. It's not a "dynamic need" (and I have no idea what that even means - the language syntax simply doesn't allow for what you want to do - accept it). It doesn't simplify authoring (Chris's @@{} splatting does that), it doesn't simplify review or maintenance.  It adds an unnecessary and BREAKING and non-obvious change between language versions. 

It's wrong. And it's bad for the language. IMO.

Reply:
>  ...the language syntax simply doesn't allow for what you want to do - accept it
> <snip>
> It adds an unnecessary and BREAKING and non-obvious change

@swngdnz I respect that you disagree.

I do not respect your continued efforts to troll on this RFC, and until that behavior changes, comments from you will be ignored.

Reply:
You assert that I "troll" because I disagree with you. I think the idea expressed by this RFC is wrong, misguided, and bad for the language.

That isn't trolling, regardless of what you think. It's an honest opinion, whether you like it or not.

As noted previously, this is a "RFC - Request For Comments". My comments may not "win", but they are as valid as anyone else's comments. 

I honestly don't care whether you ignore me or not. Others will not.

Reply:
Wrong.

> Sorry. As I have before, I simply disagree.

This is disagreeing. I respect that you disagree, as I said.

> It's wrong. And it's bad for the language. IMO.

Also disagreeing, still respected and respectful.

> ...the language syntax simply doesn't allow for what you want to do - accept it

This is trolling and unnecessary.

> It adds an unnecessary and BREAKING and non-obvious change.

This is more trolling, especially with the emphasis added given what you were responding to.

Reply:
Ok, I don't know what's going on around here. First off:

@rjmholt wrote:
> This is particularly pertinent I think because the big promise of PowerShell 7 is for Windows PowerShell 5.1 compatibility.

This is **backward** compatibility. The goal is that PowerShell 5.1 scripts will work in PowerShell 7. **This** is critical. 

However, nobody is suggesting that PowerShell 7 is going to be _the same_ as 5.1 -- many PowerShell 7 scripts **WILL NOT** work in PowerShell 5.1 -- and frankly, I don't think that should be a goal. We absolutely should fix things without new syntax when we can, but preventing new syntax just because it's new would be silly.

As far as @swngdnz -- what you're saying doesn't make sense. Both this `@` wrapping and the `@@{` hashtable syntax are new syntax which is not backward compatible. They are both addressing the exact same desire for a more readable way to pass a lot of parameters to the same command. You can prefer one over the other, but the claim that one "breaks" things (and the other doesn't) is just wrong. Neither syntax would be backward compatible, but neither would break existing scripts.

Personally, I think either would improve the authoring experience, but I like this one more.

Reply:
In re-reading my prior comment I recognise that it was actually written quite poorly, which I will blame on it being very early morning, after a bad nights sleep, whilst travelling and I hadn't properly woken up, which when compounded with the fact that I then failed to do my own due diligence and review it before posting, which would have caught many of the issues with it, forcing me to rewrite it prior to posting, but alas I was too quick to just get it posted. 

For that I apologise to anyone that read it & to those that then responded as the content of the comment was not fully aligned with my intension.

So let me clarify what my intent was.

When I said this
>The only real con that I can see is that introducing a new language feature like this will require people being made aware of the fact this only works in newer versions of PowerShell, effectively adding further breaking functionality between versions, albeit a positive break, much like the addition of DSC or Classes to the language.

I should have actually said something more like this
> The only real "con" isn't actually a con but more of an issue with making sure the right parts of awareness are there to aid the adoption & support of this new language feature.  This is the same for introducing any new language feature & will require people being made aware of the fact this only works in newer versions of PowerShell. The community needs to do what we can to ensure that script/module authors make use of the defensive coding practices that are already available including `#requires -Version` or `PowerShellVersion` in Module Manifests when they use this new feature, much like with the addition of DSC or Classes to the language. We just have to be very mindful that when people see this for the first time they will look at it and think **WTF** unless we get the supporting stuff right for user adoption. New language features aren't a con in the classical sense of the term but this really needs raised as more of a **concern** than it being a con. 

Also when I said this
> FWIW, I think this is a much needed language addition which will simplify both the authoring experience and the readability/maintenance experience, however I do feel that to state there's no known cons at this time is technically incorrect, and whilst all Language features have an implicit breaking change in language functionality meaning that those that adopt this new feature can no longer back port scripts to prior versions, which whilst we all know this is the case I can't help but feel this needs to be explicitly mentioned as a con for anyone reading this coming completely fresh to PowerShell in 2/3 years with no experience in future.

I actually intended to say something more along the lines of this

>FWIW, I think this is a much needed language addition which will simplify both the authoring experience and the readability/maintenance experience, however I do feel that to state there's no known cons at this time is technically incorrect. The reason why I think this is partially down to the fact that adopting new language features, like we have seen with DSC and Classes, breaks the user expectation of what they can do by adding something new that they couldn't do before, which is not a break like we have in typical software development practises and as such shouldn't be associated with that mindset. It is however a break, or more accurately speaking, a divergence, for those teaching or using the language, as they would need to now understand the syntax of 1 or more methods of this particular aspect of the language. So whilst all new language features have this implicit divergence to user expectations, particularly with change in language functionality meaning that those that adopt this new feature can no longer back port scripts that use that particular feature to prior versions, which whilst we all know this is the case I can't help but feel this needs to be explicitly mentioned as a loosely defined `con` for anyone reading this coming completely fresh to PowerShell in 2/3 years with no experience in future. This divergence is exactly the same as with DSC or Classes, and as such we need to ensure the adoption side is properly accounted for in changes like this.

Note that in the above I've removed almost all mentions of `breaks` - This is because as @KirkMunro rightly pointed out below there is a negative association with the terminology of `breaking change` when it comes to software development and my comments were actually less about software development but more about the learning journey for any user of this language and I should have actually used `divergence` as it is the more syntactically correct term. 

> @kilasuit's comment suggested the changes were breaking several times, but that is incorrect. My reply was not a "Defense from Comments". It was about calling out something that was incorrect though. I feel very strongly that people of influence should be careful about the wording used to discuss changes with the community, especially when talking about "breaking" functionality, which is considered very negative when it comes to software. New language features are not breaking. They are not backward compatible, because they are new language features, but they are absolutely not breaking. That is the wrong terminology to use, and I think it is important to call that out, especially in an RFC where others reading the RFC may also read the comments. This RFC proposal does not include any breaking changes.

As @KirkMunro mentioned above, yes I was technically wrong but that was due to incorrectly expressing my intended response. For that again I apologise & I hope the above amendments to my comments make of better reading than the originals which **should** now get the actual point I was intending to make, which is that the support in the user learning story needs to be **explictly** thought about and detailed in any RFC, hence I've raised #212 to add some additional items to the RFC template which should hopefully make it easier to capture these concerns within a section of the template.

However, I will say that some good points came off the back of it, as it opened some further discussion.

Like as @KirkMunro **rightfully** pointed out 

>In general though, as a community we need to push using #requires -Version in scripts and setting PowerShellVersion in module manifests as a best practice

If you have used ISE Steroids then you'd know that you get this added in via its feature set and it would scan the script and work out which language features you are are using and then add the right `#Requires -Version` statement to your script.  I am not sure about the state of this in VSCode & whether thats something that can be added into VSCode or PES but it is something that we should ensure is flagged as a new feature req over there cc @TylerLeonhardt (as not sure where this feat request should actually sit)


I'd also like to thank @rjmholt for his comment as it details just how much engineering work is needed to add new language features, something that whilst personally I was aware of the effort needed, having it detailed in the way that he has done is actually really useful for future readers of this RFC and shows just how much needs to be done to add new language features. One of the more useful parts of his response was detailing the differences in adding PS language features vs C#, as this is something that I expect is perhaps not that well documented / known about outside of the PowerShell Team and perhaps a select group of developers in the community.

One thing that I would like to say about this particular comment from @rjmholt 
>I'm just a developer, but I spend a lot of my time trying to soften discussions and prevent flame wars on GitHub now (time which I could spend enhancing PowerShell, etc.), and would prefer if we could avoid that.

Whilst I can understand that getting involved in these types of discussions can often seem to be a less productive of time than doing coding, it is always useful for the wider ecosystem to have comments from the developers on the PowerShell team, something that I would like to see more of if I am honest in these RFC's, as your comments are at least as valuable to us as the comments of the rest of us.

Also when @KirkMunro said this
>The fact that issue/challenge exists shouldn't get in the way of pushing the language forward for the better.

He was right, however I strongly feel that we need to make sure that supporting docs etc are in place to help aid adoption of new language features, partly why I have added an enhancement to the RFC template to help authors of RFC's document what they think may be needed, as this is an aspect that I see be forgotten in older enhancements, particularly the classes enhancement.

re @swngdnz point here
>  the language syntax simply doesn't allow for what you want to do

As the language stands **currently**, this is correct, however language evolves and this RFC is a part of the evolution of the language.

saying
> accept it

That's the easy way out & not of any tangible use to the topic of conversation here. Whilst it's perfectly valid for the current state of the language and it's feature set, we do not have to "accept it" and as such @KirkMunro raised this RFC to extend the language to add this functionality, functionality that I support being added.

Regarding 

>It's wrong. And it's bad for the language. IMO.

I had to re-read this 2/3 times to correctly understand the intent because initially I read it as a statement of fact followed with a statement of opinion, whereas actually it's all a statement of opinion, as confirmed by the later comment of
> I think the idea expressed by this RFC is wrong, misguided, and bad for the language

A comment that I & others here disagree with, 

In case I have not been clear, this is a language feature that I would personally like to see implemented, and like @Jaykul I prefer this to the other RFC for generalised splatting, though either would be a much needed enhancement.

Once again, apologies for my poor initial comment, I hope this one has been more useful and reads in a better light.

Reply:
I'm saddened that we have so much debate on whether it is acceptable to add new language constructs because of fragmentation concerns. The fact is, fragmentation will **always** occur and has been the case for each PS version released.

The key issue to solving fragmentation is based on the following questions:
1. How easy is to upgrade to newer version? _(partially addressed in 6.0+)_
2. Will my existing scripts work without changes? _PS 7.0 aims to make this better after 6 breakage_
3. What minimum version should I support as a publisher of Modules?

So does this RFC impact any of those questions. Yes, The 3rd question. As a Module author that publishes to PS Gallery, I need to use syntax that works across several versions so I will not be able to opt-in to this feature. That sucks, but it is what it is. We should certainly do all that we can to make it easier for Module authors to use the right syntax for the broadest community support. PS 2.0 support held back a lot of people, including Chocolatey, for this very reason.

Now should that block me from my own personal scripts or modules that are internal? No, I have more choice. I personally now only write for PS 5.1 support since all of my servers are using PS 5.1 (finally). The fact that 3rd parties are held back on an old version, doesn't hold me back. I can and want to write more efficient scripts and cleaner code. 

I should be allowed to have that choice rather than being held back artificially by never advancing new features because of fragmentation concerns.

Reply:
It's incorrect to call the continuing development of a language divergence or fragmentation.

There's none of that going on. There's only one language implementation. This isn't a fork -- it's the one and only official language. Yes, we're adding new features -- and we must continue to do so if you want to stay relevant. We are not breaking old things. If you want to support the old engine, you simply don't use the new features, and 

# EVERYTHING just works, same as always!

Look at Javascript's ECMAScript 6 by comparison. The syntax improvements there are enormous compared to these tiny changes, but nobody's calling that a divergence or a fragmentation or anything like that: the old code continues working, and people who need the new features use the new features, (and perhaps use conversion tools to support the old engine, if it's worth it).

Reply:
New versions always create fragmentation. Product fragmentation is defined as several different versions of a product with each having significant market share. Windows XP vs 7 vs 10 is a great example of this. In our world, PS 2.0 vs 3.0 vs 5.1 vs 6.x is more comparable and even by Windows Server since a lot of Windows cmdlets are tied to the Windows release.

Most of the negative comments sounds to me to be concerns about the market being so fragmented that using a new language construct won't get used. My point is that unless a feature inhibits upgrades (breaking changes) then we shouldn't care about the fragmentation issue with regards to this RFC. So in that point, I agree with you @Jaykul and @KirkMunro. I'm just making the point in a different way.

How we solve fragmentation issues within the PowerShell community are really separate issues and RFCs. Some of which I've already seen discussions around. 

Either way, these issues have zero bearing on whether this RFC should be approved or not. The RFC needs to be approved on its merit and whether or not the implementation is sound for solving a problem, and yes I would say better multi-line continuance is a real problem to solve that this solves in an elegant way.

Reply:
First, let me say I appreciate the passion and interest in this discussion.  I do ask that we try to keep the discussion civil and professional.  Not everyone will be pleased by some new feature and that's ok.  We should accept all feedback and try to distill their meta-point as English is not everyone's first language and different folks have different communication styles (you should see some of the early sarcastic tweets I made early on that got me into social media hot water...).

As others have stated, PowerShell will continue to evolve and an explicit goal is to maintain *backwards* compatibility (there are exceptions, of course), but *forward* compatibility is not a goal although a consideration (although it's more of a "if we do this, will it prevent other innovation in the future").  Any new cmdlet, parameter, language addition will not work on older versions of PowerShell by definition.  We now have _Experimental Features_ to allow experiencing new features in code rather than just reading it in a spec to evaluate the design.  I like @KirkMunro's proposal on the concept of [Optional Features](https://github.com/PowerShell/PowerShell-RFC/pull/187/files) as a way for users to opt-in or opt-out of changes to PowerShell that may mitigate some of these concerns (although it certainly has its own cons).  Also keep in mind:

1. a "feature" of Windows PowerShell 5.1 being complete is that it will not be getting new features so you can rely on it's current behavior for your mission critical scripts for the foreseeable future
2. PowerShell Core 6 and PowerShell 7 run side-by-side so you can target specific scripts for a specific version of PowerShell.  @KirkMunro has a [RFC](https://github.com/PowerShell/PowerShell-RFC/pull/202/files) proposing how to make this easier

Reply:
I'm marking this portion of the conversation as resolved. I believe everything that came up that should be addressed in the RFC is included in the RFC.

Full disclosure: I did not add considerations or cons for the following:

* a language syntax change will not work in older versions of PowerShell;

    This is by definition, because the PowerShell Team does not backport language syntax changes to older versions of PowerShell. Further, none of the RFCs that propose syntax changes call this out.

* changing language syntax can result in a lot of (hard) work;

    This information, while useful, should be described in documentation in the PowerShell and/or PowerShell-RFC repositories.

For these two issues, I shared some of what I think would make for an improved RFC experience for RFC authors and readers alike in [a comment to another PR in this repo](https://github.com/PowerShell/PowerShell-RFC/pull/212#issuecomment-511247116). If you have thoughts in that area and like/dislike, let's continue that discussion over there, or in a new issue in this repo.

If there is another detail in this part of the conversation that I missed among all of the chatter that should be included in this RFC, please call that out so that I can make sure it is considered.